R/formulas.R

```diff
@@ -0,0 +1,173 @@
+#'
+#' \code{fit_with()} is a pipe-friendly tool that applies a list of
+#' formulas to a fitting function such as \code{\link[stats]{lm}()}.
+#' The list of formulas is typically created with \code{\link{formulas}}().
+#'
+#' Assumes that \code{.f} takes the formula either as first argument
+#' or as second argument if the first argument is \code{data}.  Most
+#' fitting functions should fit these requirements.
+#' @param data A dataset used to fit the models.
+#' @param .f A fitting function such as \code{\link[stats]{lm}()},
+#'   \code{\link[lme4]{lmer}}() or \code{\link[rstanarm]{stan_glmer}()}.
+#' @param .formulas A list of formulas specifying a model.
+#' @param ... Additional arguments passed on to \code{.f}
+#' @seealso \code{\link{formulas}}()
+#' @export
+#' @examples
+#' # fit_with() is typically used with formulas().
+#' disp_fits <- mtcars %>% fit_with(lm, formulas(~disp,
+#'   additive = ~drat + cyl,
+#'   interaction = ~drat * cyl,
+#'   full = add_predictors(interaction, ~am, ~vs)
+#' ))
+#'
+#' # The list of fitted models is named after the names of the list of
+#' # formulas:
+#' disp_fits$full
+#'
+#' # Additional arguments are passed on to .f
+#' mtcars %>% fit_with(glm, list(am ~ disp), family = binomial)
+fit_with <- function(data, .f, .formulas, ...) {
+  args <- list(...)
+
+  # Shadowing data to avoid getting the whole raw SEXP structure in
+  # the calls captured by the fitting function (which gets displayed
+  # in print methods)
+  args$data <- quote(data)
+
+  map(.formulas, function(formula) {
+    do.call(".f", c(list(formula), args))
+  })
+}
+
+#' Create a list of formulas
+#'
+#' \code{formulas()} creates a list of two-sided formulas by merging a
+#' unique left-hand side to a list of right-hand sides.
+#' @param .response A one-sided formula used as the left-hand side of
+#'   all resulting formulas.
+#' @param ... List of formulas whose right-hand sides will be merged
+#'   to \code{.response}.
+#' @export
+#' @examples
+#' # Provide named arguments to create a named list of formulas:
+#' models <- formulas(~lhs,
+#'   additive = ~var1 + var2,
+#'   interaction = ~var1 * var2
+#' )
+#' models$additive
+#'
+#' # The formulas are created sequentially, so that you can refer to
+#' # previously created formulas:
+#' formulas(~lhs,
+#'   linear = ~var1 + var2,
+#'   hierarchical = add_predictors(linear, ~(1 | group))
+#' )
+formulas <- function(.response, ...) {
+  formulas <- tibble::lst(...)
+  validate_formulas(.response, formulas)
+  map(formulas, set_lhs, .response)
+}
+
+#' @rdname formulas
+#' @export
+formulae <- formulas
+
+validate_formulas <- function(response, formulas) {
+  if (!is_formula(response)) {
+    stop(".response should be a formula", call. = FALSE)
+  }
+
+  if (!length(formulas)) {
+    stop("No formula provided", call. = FALSE)
+  }
+
+  purrr::walk(formulas, function(f) {
+    if (!is_formula(f)) {
+      stop("'...' should only contain formulas", call. = FALSE)
```

Comment:
I think "contain only" is better English, but I couldn't say why and I might be wrong